pkg/discovery/grpcresolver/grpc_resolver_test.go

```diff
@@ -0,0 +1,163 @@
+	for _, s := range t.servers {
+		s.Stop()
+	}
+}
+
+// ErrorDiscoverer yields a discoverer that returns error
+type ErrorDiscoverer []string
+
+// Instances implements Discoverer.
+func (d ErrorDiscoverer) Instances() ([]string, error) {
+	return nil, errors.New("error discoverer always return error")
+}
+
+func startTestServers(t *testing.T, count int) *test {
+	testInstance := &test{}
+	for i := 0; i < count; i++ {
+		lis, err := net.Listen("tcp", "localhost:0")
+		assert.NoError(t, err, "failed to listen on tcp")
+		s := grpc.NewServer()
+		grpctest.RegisterTestServiceServer(s, &testServer{})
+		testInstance.servers = append(testInstance.servers, s)
+		testInstance.addresses = append(testInstance.addresses, lis.Addr().String())
+
+		go func(s *grpc.Server, l net.Listener) {
+			s.Serve(l)
+		}(s, lis)
+	}
+
+	return testInstance
+}
+
+func makeSureConnectionsUp(t *testing.T, count int, testc grpctest.TestServiceClient) {
+	var p peer.Peer
+	addrs := make(map[string]struct{})
+	// Make sure connections to all servers are up.
+	for si := 0; si < count; si++ {
+		connected := false
+		for i := 0; i < 1000; i++ {
+			_, err := testc.EmptyCall(context.Background(), &grpctest.Empty{}, grpc.Peer(&p))
+			if err != nil {
+				continue
+			}
+			if _, ok := addrs[p.Addr.String()]; !ok {
+				addrs[p.Addr.String()] = struct{}{}
+				connected = true
+				break
+			}
+			time.Sleep(time.Millisecond * 10)
+		}
+		assert.True(t, connected, "Connection was still not up")
+	}
+}
+
+func assertRoundRobinCall(t *testing.T, connections int, testc grpctest.TestServiceClient) {
+	addrs := make(map[string]struct{})
+	var p peer.Peer
+	for i := 0; i < connections; i++ {
+		_, err := testc.EmptyCall(context.Background(), &grpctest.Empty{}, grpc.Peer(&p))
+		assert.NoError(t, err)
+		addrs[p.Addr.String()] = struct{}{}
+	}
+	assert.Len(t, addrs, connections, "must call each of the servers once")
+}
+
+func TestErrorDiscoverer(t *testing.T) {
+	notifier := &discovery.Dispatcher{}
+	discoverer := ErrorDiscoverer{}
+	r := New(notifier, discoverer, zap.NewNop(), 2)
+	_, err := r.Build(resolver.Target{}, nil, resolver.BuildOption{})
+	assert.Error(t, err)
+}
+
+func TestGRPCResolverRoundRobin(t *testing.T) {
+	backendCount := 5
+
+	test := startTestServers(t, backendCount)
+	defer test.cleanup()
+
+	notifier := &discovery.Dispatcher{}
+	discoverer := discovery.FixedDiscoverer{}
+
+	for _, minPeers := range []int{3, 5, 7} {
+		res := New(notifier, discoverer, zap.NewNop(), minPeers)
+		defer resolver.UnregisterForTesting(res.Scheme())
+
+		cc, err := grpc.Dial(res.Scheme()+":///round_robin", grpc.WithInsecure(), grpc.WithBalancerName(roundrobin.Name))
+		defer cc.Close()
+		assert.NoError(t, err, "could not dial using resolver's scheme")
+
+		testc := grpctest.NewTestServiceClient(cc)
+
+		notifier.Notify(test.addresses)
+
+		connections := minPeers
+		if connections > backendCount {
+			connections = backendCount
+		}
+
+		makeSureConnectionsUp(t, connections, testc)
```

Comment:
this should be called with `backendCount`, not `connections`. Or even better - it should be outside of the loop, since you're not restarting servers (which is why I suggested making it part of startTestServers, since the expectation of that function is that "servers are ready"). And you don't need to use testc with the load balancer for this health check.